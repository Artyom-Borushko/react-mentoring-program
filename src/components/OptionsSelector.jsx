import React, { useState, useRef, useEffect } from "react";
import '../styles/optionsSelector.css';

export default function OptionsSelector({movieDropdownOptions, onDropdownSelection}) {
    const [isOptionsDropdownVisible, setOptionsDropdownVisible] = useState(false);
    const dropdownRef = useRef(null);

    const toggleDropdown = () => {
        setOptionsDropdownVisible(!isOptionsDropdownVisible);
    };

    useEffect(() => {
        const handleClickOutside = (event) => {
            if (dropdownRef.current && !dropdownRef.current.contains(event.target)) {
                setOptionsDropdownVisible(false);
            }
        };

        document.addEventListener("mousedown", handleClickOutside);

        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        };
    }, []);

    return (
        <div ref={dropdownRef}>
            <img
                src={'images/options-selector/threeDotsExpandOptionsIcon.png'}
                alt='expand options'
                className={'movie-expand-options-icon'}
                aria-label={'expand options'}
                onClick={toggleDropdown}
            />

            {isOptionsDropdownVisible && (
                <div className="dropdown">
                    <img
                        src="/images/options-selector/whiteCrossSign.png"
                        alt="white cross"
                        className={'close-dropdown'}
                        aria-label={'close dropdown'}
                        onClick={toggleDropdown}
                    />
                    {movieDropdownOptions.map((dropdownOption, index) => {
                        return (
                            <div
                                key={index}
                                className={'dropdown-option'}
                                onClick={onDropdownSelection}
                            >
                                {dropdownOption}
                            </div>
                        )
                    })}
                </div>
            )}
        </div>
    );
};
