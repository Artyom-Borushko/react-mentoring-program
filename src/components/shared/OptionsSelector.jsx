import React, { useState, useRef, useEffect } from "react";
import './optionsSelector.css';

export default function OptionsSelector({dropdownOptions, onDropdownSelection, controlImageSrc, controlSource = ''}) {
    const [isOptionsDropdownVisible, setOptionsDropdownVisible] = useState(false);
    const dropdownRef = useRef(null);

    const toggleDropdown = () => {
        setOptionsDropdownVisible(!isOptionsDropdownVisible);
    };

    const handleDropdownSelect = (event) => {
        toggleDropdown();
        onDropdownSelection(event);
    };

    useEffect(() => {
        const handleClickOutside = (event) => {
            if (dropdownRef.current && !dropdownRef.current.contains(event.target)) {
                setOptionsDropdownVisible(false);
            }
        };

        document.addEventListener("mousedown", handleClickOutside);

        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        };
    }, []);

    return (
        <div ref={dropdownRef}>
            <img
                src={controlImageSrc}
                alt='expand options'
                className={`movie-expand-options-icon ${controlSource}`}
                aria-label={'expand options'}
                onClick={toggleDropdown}
            />

            {isOptionsDropdownVisible && (
                <div className="dropdown">
                    <img
                        src="/images/options-selector/whiteCrossSign.png"
                        alt="white cross"
                        className={'close-dropdown'}
                        aria-label={'close dropdown'}
                        onClick={toggleDropdown}
                    />
                    {dropdownOptions.map((dropdownOption, index) => {
                        return (
                            <div
                                key={index}
                                className={'dropdown-option'}
                                onClick={handleDropdownSelect}
                            >
                                {dropdownOption}
                            </div>
                        )
                    })}
                </div>
            )}
        </div>
    );
};
