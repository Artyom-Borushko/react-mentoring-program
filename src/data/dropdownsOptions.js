export const movieDropdownOptions = ['edit', 'delete'];

export const moviesSortDropdownOptions = ['release date', 'title'];
