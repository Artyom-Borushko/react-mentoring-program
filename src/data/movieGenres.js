
export const listOfGenres = ['all', 'documentary', 'comedy', 'horror', 'crime'];
