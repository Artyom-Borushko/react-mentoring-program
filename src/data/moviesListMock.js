export const listOfMoviesMock = [
    { src: '/images/movies-list/pulpFiction.png', title: "Pulp Fiction", releaseYear: "2004", relevantGenres: ['action', 'crime'], rating: '9.2', duration: '2h 22min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId1" },
    { src: '/images/movies-list/bohemianRhapsody.png', title: "Bohemian Rhapsody", releaseYear: "2003", relevantGenres: ['documentary', 'drama'], rating: '8.4', duration: '1h 27min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId2" },
    { src: '/images/movies-list/killBillVol2.png', title: "Kill Bill: Vol 2", releaseYear: "1994", relevantGenres: ['comedy', 'crime', 'oscar'], rating: '6.2', duration: '3h 12min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId3" },
    { src: '/images/movies-list/avengersWarOfInfinity.png', title: "Avengers: War of Infinity", releaseYear: "2004", relevantGenres: ['action', 'adventure'], rating: '8.2', duration: '0h 56min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId4" },
    { src: '/images/movies-list/inception.png', title: "Inception", releaseYear: "2003", relevantGenres: ['action', 'adventure', 'horror'], rating: '10.0', duration: '1h 39min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId5" },
    { src: '/images/movies-list/reservoirDogs.png', title: "Reservoir dogs", releaseYear: "1994", relevantGenres: ['crime', 'oscar'], rating: '7.7', duration: '2h 00min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId6" },
];
