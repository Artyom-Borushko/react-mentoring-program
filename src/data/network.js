export default 'http://localhost:4000/movies';
