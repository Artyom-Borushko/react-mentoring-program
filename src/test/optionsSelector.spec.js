import {render, screen} from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import '@testing-library/jest-dom';
import OptionsSelector from "../components/OptionsSelector";
import {movieDropdownOptions} from "../data/dropdownsOptions";

describe("OptionsSelector component tests", () => {

    it("renders OptionsSelector component", () => {
        const {container} = render(<OptionsSelector
            movieDropdownOptions={movieDropdownOptions}
        />);

        expect(container).toMatchSnapshot();
    });

    it("does not displays options dropdown by default", () => {
        render(<OptionsSelector
            movieDropdownOptions={movieDropdownOptions}
        />);

        expect(screen.queryByText(movieDropdownOptions[0])).not.toBeInTheDocument();
    });

    it("displays options dropdown after click on options selector icon", async () => {
        render(<OptionsSelector
            movieDropdownOptions={movieDropdownOptions}
        />);

        await userEvent.click(screen.getByRole('img', {name: 'expand options'}));

        expect(screen.getByText(movieDropdownOptions[0])).toBeInTheDocument();
    });

    it("closes options dropdown after click in any place", async () => {
        render(<OptionsSelector
            movieDropdownOptions={movieDropdownOptions}
        />);

        await userEvent.click(screen.getByRole('img', {name: 'expand options'}));
        await userEvent.click(document.body);

        expect(screen.queryByText(movieDropdownOptions[0])).not.toBeInTheDocument();
    });

    it("closes options dropdown after click in cross icon", async () => {
        render(<OptionsSelector
            movieDropdownOptions={movieDropdownOptions}
        />);

        await userEvent.click(screen.getByRole('img', {name: 'expand options'}));
        await userEvent.click(screen.getByRole('img', {name: 'close dropdown'}));

        expect(screen.queryByText(movieDropdownOptions[0])).not.toBeInTheDocument();
    });

    it("does not close options dropdown after click in dropdown element", async () => {
        render(<OptionsSelector
            movieDropdownOptions={movieDropdownOptions}
        />);

        await userEvent.click(screen.getByRole('img', {name: 'expand options'}));
        await userEvent.click(screen.getByText(movieDropdownOptions[0]));

        expect(screen.getByText(movieDropdownOptions[0])).toBeInTheDocument();
    });

    it("event listener is called when selecting option from dropdown", async () => {
        const mockClickHandler = jest.fn();
        render(<OptionsSelector
            movieDropdownOptions={movieDropdownOptions}
            onDropdownSelection={mockClickHandler}
        />);

        await userEvent.click(screen.getByRole('img', {name: 'expand options'}));
        await userEvent.click(screen.getByText(movieDropdownOptions[1]));

        expect(mockClickHandler).toHaveBeenCalledTimes(1);
    });
});
