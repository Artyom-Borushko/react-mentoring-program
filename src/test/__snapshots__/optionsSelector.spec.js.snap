// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsSelector component tests renders OptionsSelector component 1`] = `
<div>
  <div>
    <img
      alt="expand options"
      aria-label="expand options"
      class="movie-expand-options-icon"
      src="images/options-selector/threeDotsExpandOptionsIcon.png"
    />
  </div>
</div>
`;
